query/parse.go

```diff
@@ -72,11 +72,21 @@ func isSpace(c byte) bool {
 	return c == ' ' || c == '\t'
 }
 
+// globally tracking parentheses
+var parenBalance int
```

Comment:
this is a race condition since Parse can be called concurrently. However, I think you can just change line 81 below in Parse to be
```
parenBalance := 0
```

then you can remove the global?